File: apps/grokking-ml-fsharp-lib/src/GrokkingMlLib/Nb11UniversityAdmissions.fs
/// 原著ノートブック #11 `Chapter_09_Decision_Trees/University_Admissions.ipynb`。
///
/// 大学院の入学審査データ 400 件から、合格するかどうかを決定木で当てる。
/// 原著は **同じデータで木の大きさを変えて、過学習の様子を見せる** 構成になっている。
///
/// [#10](Nb10DecisionTreeBoundary.fs) の CART を、
/// 任意の特徴量・葉の最小件数・分割の最小件数に対応できるように一般化して使う。
/// 原著が掛ける 3 つの制限（深さ・葉の件数・分割の件数）をすべて再現する。
module GrokkingMlLib.Nb11UniversityAdmissions

open Deedle
open GrokkingMlLib.Nb08GiniEntropy

/// 原著が合格とみなす基準
[<Literal>]
let AdmissionThreshold = 0.75

/// 特徴量の列。原著の順序をそのまま保つ
let featureNames =
    [ "GRE Score"; "TOEFL Score"; "University Rating"; "SOP"; "LOR"; "CGPA"; "Research" ]

/// 2 特徴量だけで学習するときに使う列
let examFeatures = [ "GRE Score"; "TOEFL Score" ]

/// 読み込んだデータ。列名 -> 値と、合否ラベル
type Admissions =
    { Columns: Map<string, float[]>
      Admitted: string[] }

    member this.Size = this.Admitted.Length

    /// 指定した列だけを行ベクトルの配列にする
    member this.RowsOf(names: string list) =
        Array.init this.Size (fun row -> names |> List.map (fun name -> this.Columns.[name].[row]) |> Array.ofList)

/// 決定木。#09・#10 と同じ判別共用体
type Tree =
    | Leaf of prediction: string
    | Node of feature: string * threshold: float * left: Tree * right: Tree

/// 入学審査データを読み込み、合否のラベルを付ける。
///
/// 原著は `Chance of Admit`（合格確率）を 0.75 で切って 2 値にし、元の列を落としている。
let loadData () =
    let frame = Datasets.loadFrame "Admission_Predict.csv"

    let columns =
        featureNames
        |> List.map (fun name -> name, frame.GetColumn<float>(name) |> Series.values |> Array.ofSeq)
        |> Map.ofList

    let admitted =
        frame.GetColumn<float>("Chance of Admit")
        |> Series.values
        |> Seq.map (fun chance -> if chance >= AdmissionThreshold then "true" else "false")
        |> Array.ofSeq

    { Columns = columns; Admitted = admitted }

let private candidateThresholds (values: float[]) =
    values |> Array.distinct |> Array.sort |> Array.pairwise |> Array.map (fun (a, b) -> (a + b) / 2.0)

/// 葉の予測クラス。もっとも多いクラスを返す。
///
/// **同数のときはクラス名の小さいほうを選ぶ。** scikit-learn は `np.argmax` を
/// クラス順に並べた件数配列へ適用するので、同数なら先頭のクラス（ここでは
/// `"false"`）が選ばれる。原著が「白い（中立な）葉は False になる」と
/// コメントしているのはこの挙動である。出現順で選ぶと結果が変わる。
let private majority (labels: string[]) =
    labels
    |> Array.countBy id
    |> Array.sortBy (fun (label, count) -> -count, label)
    |> Array.head
    |> fst

/// 木を育てるときの制限。原著の 3 つの引数にそのまま対応する
type Limits =
    { /// `max_depth`。根を深さ 0 と数える
      MaxDepth: int
      /// `min_samples_leaf`。葉に残る件数の下限
      MinSamplesLeaf: int
      /// `min_samples_split`。分割してよい節の件数の下限
      MinSamplesSplit: int }

/// 制限なし
let unlimited =
    { MaxDepth = 100
      MinSamplesLeaf = 1
      MinSamplesSplit = 2 }

let private weightedGini (rows: float[][]) (labels: string[]) column threshold minSamplesLeaf =
    let indices = Array.init rows.Length id
    let leftIndices = indices |> Array.filter (fun i -> rows.[i].[column] <= threshold)
    let rightIndices = indices |> Array.filter (fun i -> rows.[i].[column] > threshold)

    if leftIndices.Length < minSamplesLeaf || rightIndices.Length < minSamplesLeaf then
        None
    else
        let labelsOf (indices: int[]) = indices |> Array.map (fun i -> labels.[i]) |> List.ofArray
        let leftLabels = labelsOf leftIndices
        let rightLabels = labelsOf rightIndices
        let n = float rows.Length

        let impurity =
            (gini leftLabels * float leftLabels.Length + gini rightLabels * float rightLabels.Length) / n

        Some(impurity, leftIndices, rightIndices)

/// CART を再帰的に構築する。原著の 3 つの制限をすべて見る
let rec private build (names: string list) limits depth (rows: float[][]) (labels: string[]) =
    let isPure = (labels |> Array.distinct |> Array.length) <= 1

    if isPure || depth >= limits.MaxDepth || labels.Length < limits.MinSamplesSplit then
        Leaf(majority labels)
    else
        let best =
            names
            |> List.indexed
            |> List.collect (fun (column, name) ->
                rows
                |> Array.map (fun row -> row.[column])
                |> candidateThresholds
                |> Array.choose (fun threshold ->
                    weightedGini rows labels column threshold limits.MinSamplesLeaf
                    |> Option.map (fun (impurity, left, right) -> impurity, name, threshold, left, right))
                |> List.ofArray)
            |> List.sortBy (fun (impurity, _, _, _, _) -> impurity)
            |> List.tryHead

        match best with
        | None -> Leaf(majority labels)
        | Some(_, name, threshold, leftIndices, rightIndices) ->
            let subset (indices: int[]) =
                indices |> Array.map (fun i -> rows.[i]), indices |> Array.map (fun i -> labels.[i])

            let leftRows, leftLabels = subset leftIndices
            let rightRows, rightLabels = subset rightIndices

            Node(
                name,
                threshold,
                build names limits (depth + 1) leftRows leftLabels,
                build names limits (depth + 1) rightRows rightLabels
            )

/// 決定木を学習する
let fit (data: Admissions) (names: string list) (limits: Limits) =
    build names limits 0 (data.RowsOf names) data.Admitted

/// 制限なしの木。訓練データを完全に覚えてしまう
let fitFull (data: Admissions) = fit data featureNames unlimited

/// 原著が「過学習しない小さい木」として作る設定
let fitSmaller (data: Admissions) =
    fit data featureNames { MaxDepth = 3; MinSamplesLeaf = 10; MinSamplesSplit = 10 }

/// GRE と TOEFL の 2 特徴量だけで学習する
let fitExams (data: Admissions) (maxDepth: int) =
    fit data examFeatures { unlimited with MaxDepth = maxDepth }

/// 1 行を予測する
let rec predict (tree: Tree) (names: string list) (row: float[]) =
    match tree with
    | Leaf prediction -> prediction
    | Node(feature, threshold, left, right) ->
        let column = List.findIndex ((=) feature) names

        if row.[column] <= threshold then
            predict left names row
        else
            predict right names row

/// 学習データに対する正解率
let accuracy (tree: Tree) (data: Admissions) (names: string list) =
    let rows = data.RowsOf names

    let correct =
        Array.map2 (fun row label -> predict tree names row = label) rows data.Admitted
        |> Array.filter id
        |> Array.length

    float correct / float data.Size

/// 1 人ぶんの出願情報から合否を予測する
let predictApplicant (tree: Tree) (values: float list) =
    predict tree featureNames (Array.ofList values) = "true"

/// 分割に使われた条件を、根から深さ優先で並べる
let rec splitConditions (tree: Tree) =
    match tree with
    | Leaf _ -> []
    | Node(feature, threshold, left, right) ->
        (feature, threshold) :: splitConditions left @ splitConditions right

/// 節と葉を合わせた数
let rec nodeCount (tree: Tree) =
    match tree with
    | Leaf _ -> 1
    | Node(_, _, left, right) -> 1 + nodeCount left + nodeCount right

File: apps/grokking-ml-python-lib/src/grokking_ml_lib/nb11_university_admissions.py
"""原著ノートブック #11 `Chapter_09_Decision_Trees/University_Admissions.ipynb`。

大学院の入学審査データ 400 件から、合格するかどうかを決定木で当てる。
原著は **同じデータで木の大きさを変えて、過学習の様子を見せる** 構成になっている。

1. 制限なしの木 — 訓練データを 100% 当てる（明らかに過学習）
2. 深さ 3・葉の最小件数 10 の木 — 88.5% に落ちるが、木が読める大きさになる
3. 2 特徴量（GRE と TOEFL）だけの木を深さ 1・2・無制限で比べる

[#10](nb10.md) が 12 点の作り物だったのに対し、こちらは実データである。
**訓練データの正解率が 1.0 になっても嬉しくない** ことがはっきり出る。
"""

from __future__ import annotations

import pandas as pd
from sklearn.tree import DecisionTreeClassifier

from grokking_ml_lib.datasets import dataset_path

#: 原著が合格とみなす基準
ADMISSION_THRESHOLD = 0.75

#: 2 特徴量だけで学習するときに使う列
EXAM_FEATURES = ["GRE Score", "TOEFL Score"]


def load_data() -> pd.DataFrame:
    """入学審査データを読み込み、合否のラベルを付ける。

    原著は `Chance of Admit`（合格確率）を 0.75 で切って 2 値にし、
    元の列を落としている。`index_col=0` で `Serial No.` を索引にする。
    """
    data = pd.read_csv(dataset_path("Admission_Predict.csv"), index_col=0)
    data["Admitted"] = data["Chance of Admit"] >= ADMISSION_THRESHOLD
    return data.drop(["Chance of Admit"], axis=1)


def features(data: pd.DataFrame) -> pd.DataFrame:
    return data.drop(["Admitted"], axis=1)


def labels(data: pd.DataFrame) -> pd.Series:
    return data["Admitted"]


def fit_full(data: pd.DataFrame, random_state: int | None = 0) -> DecisionTreeClassifier:
    """制限なしの木。訓練データを完全に覚えてしまう。"""
    return DecisionTreeClassifier(random_state=random_state).fit(
        features(data), labels(data)
    )


def fit_smaller(data: pd.DataFrame, random_state: int | None = 0) -> DecisionTreeClassifier:
    """原著が「過学習しない小さい木」として作る設定。

    3 つの制限を同時に掛けている。
    - `max_depth=3` — 木の深さ
    - `min_samples_leaf=10` — 葉に残る件数の下限
    - `min_samples_split=10` — 分割してよい節の件数の下限
    """
    return DecisionTreeClassifier(
        max_depth=3,
        min_samples_leaf=10,
        min_samples_split=10,
        random_state=random_state,
    ).fit(features(data), labels(data))


def fit_exams(
    data: pd.DataFrame, max_depth: int | None, random_state: int | None = 0
) -> DecisionTreeClassifier:
    """GRE と TOEFL の 2 特徴量だけで学習する。境界を 2 次元で描けるようにするため。"""
    return DecisionTreeClassifier(max_depth=max_depth, random_state=random_state).fit(
        data[EXAM_FEATURES], labels(data)
    )


def predict_applicant(
    model: DecisionTreeClassifier, data: pd.DataFrame, values: list[float]
) -> bool:
    """1 人ぶんの出願情報から合否を予測する。

    列名つきの 1 行に組み立ててから渡す。生のリストを渡すと
    scikit-learn が「学習時は列名があったのに」と警告を出す。
    """
    row = pd.DataFrame([values], columns=features(data).columns)
    return bool(model.predict(row)[0])


def split_conditions(
    model: DecisionTreeClassifier, feature_names: list[str]
) -> list[tuple[str, float]]:
    """木のなかで実際に分割に使われた条件を、節の順に並べる。"""
    tree = model.tree_
    return [
        (feature_names[feature], float(threshold))
        for feature, threshold in zip(tree.feature, tree.threshold)
        if feature >= 0
    ]
